idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,315 @@
+            // members.
+            val commands = analyse(classOrObject) {
+                insertedBlocks.mapNotNull { block ->
+                    val declarations = block.declarations.mapNotNull { it.element }
+                    val first = declarations.firstOrNull() ?: return@mapNotNull null
+                    val last = declarations.last()
+                    collectPossibleReferenceShortenings(first.containingKtFile, TextRange(first.startOffset, last.endOffset))
+                }
+            }
+            commands to insertedBlocks
+        }
+        runWriteAction {
+            commands.forEach { it.invokeShortening() }
+            val project = classOrObject.project
+            val codeStyleManager = CodeStyleManager.getInstance(project)
+            insertedBlocks.forEach { it.declarations.forEach { it.element?.let { element -> codeStyleManager.reformat(element) } } }
+            insertedBlocks.firstOrNull()?.declarations?.firstNotNullResult { it.element }?.let {
+                moveCaretIntoGeneratedElement(editor, it)
+            }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun KtAnalysisSession.generateMembers(
+        editor: Editor,
+        currentClass: KtClassOrObject,
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ): List<InsertedBlock> {
+        if (selectedElements.isEmpty()) return emptyList()
+        val selectedMemberSymbolsAndGeneratedPsi: Map<KtCallableSymbol, KtCallableDeclaration> = buildMap {
+            for (it in selectedElements) {
+                put(it.symbol, generateMember(currentClass.project, it, currentClass, copyDoc))
+            }
+        }
+
+        val classBody = currentClass.body
+        val offset = editor.caretModel.offset
+
+        // Insert members at the cursor position if the cursor is within the class body.
+        if (classBody == null || isCursorInsideClassBodyExcludingBraces(classBody, offset)) {
+            return runWriteAction {
+                listOf(InsertedBlock(insertMembersAfter(editor, currentClass, selectedMemberSymbolsAndGeneratedPsi.values)))
+            }
+        }
+
+        // Insert members at positions such that the result aligns with ordering of members in super types.
+        val orderedMembers = getMembersOrderedByRelativePositionsInSuperTypes(currentClass, selectedMemberSymbolsAndGeneratedPsi)
+        return insertMembersAccordingToPreferredOrder(orderedMembers, classBody.lBrace, editor, currentClass)
+    }
+
+    private fun isCursorInsideClassBodyExcludingBraces(classBody: KtClassBody, offset: Int): Boolean {
+        return classBody.textRange.contains(offset)
+                && classBody.lBrace?.textRange?.contains(offset) == false
+                && classBody.rBrace?.textRange?.contains(offset) == false
+    }
+
+    /**
+     * Given a class and some stub implementation of overridden members, output all the callable members in the desired order. For example,
```

Comment:
FE1.0 logic is at https://github.com/JetBrains/kotlin/blob/e455e926aad35a792cc86c01e288a0cd3b8122d9/idea/idea-core/src/org/jetbrains/kotlin/idea/core/overrideImplement/OverrideImplementMembersHandler.kt#L117. It's much shorter since it applies chanes to the PSI directly and retrieve new descriptors after the change after every operation.